rsocket-core/src/main/java/io/rsocket/metadata/CompositeMetadata.java

```diff
@@ -0,0 +1,309 @@
+     * {@link ByteBuf#readRetainedSlice(int) retained} if needed.
+     *
+     * @param buffer the buffer to decode
+     * @param retainMetadataSlices should each slide be retained when read from the original buffer?
+     * @return the decoded {@link Entry}
+     */
+    public static Entry decodeIncrementally(ByteBuf buffer, boolean retainMetadataSlices) {
+        Object[] entry = CompositeMetadataFlyweight.decodeNext(buffer, retainMetadataSlices);
+        Object mime = entry[0];
+        ByteBuf buf = (ByteBuf) entry[1];
+        if (mime instanceof WellKnownMimeType) {
+            return new CompressedTypeEntry((WellKnownMimeType) mime, buf);
+        }
+        if (mime instanceof Byte) {
+            return new UnknownCompressedTypeEntry((Byte) mime, buf);
+        }
+        return new CustomTypeEntry((String) mime, buf);
+    }
+
+    /**
+     * Encode a {@link CompositeMetadata} into a new {@link CompositeByteBuf}.
+     * <p>
+     * This method moves the buffer's {@link ByteBuf#writerIndex()}.
+     * It uses the existing content {@link ByteBuf} of each {@link Entry}, but allocates a new buffer for each metadata
+     * header using the provided {@link ByteBufAllocator}.
+     *
+     * @param allocator the {@link ByteBufAllocator} to use when a new buffer is needed
+     * @param metadata the {@link CompositeMetadata} to encode
+     * @return a {@link CompositeByteBuf} that represents the {@link CompositeMetadata}
+     */
+    public static CompositeByteBuf encode(ByteBufAllocator allocator, CompositeMetadata metadata) {
+        CompositeByteBuf compositeMetadataBuffer = allocator.compositeBuffer();
+        for (Entry entry : metadata.getAll()) {
+            encodeIncrementally(allocator, compositeMetadataBuffer, entry);
+        }
+        return compositeMetadataBuffer;
+    }
+
+    /**
+     * Incrementally encode a {@link CompositeMetadata} by encoding a provided {@link Entry} into a pre-existing
+     * {@link CompositeByteBuf} (which can be empty if it is the first entry that is being encoded).
+     * <p>
+     * This method moves the composite buffer's {@link ByteBuf#writerIndex()}, and internally allocates one buffer for
+     * the metadata header using the provided {@link ByteBufAllocator}.
+     * <p>
+     * If the mime type is either a {@link WellKnownMimeType} or a {@link WellKnownMimeType#UNKNOWN_RESERVED_MIME_TYPE},
+     * it is compressed using the identifier. Otherwise, the {@link String} length + mime type are encoded.
+     *
+     * @param allocator the {@link ByteBufAllocator} to use to encode the metadata headers (mime type id/length+string, metadata length)
+     * @param compositeMetadataBuffer the composite buffer to which this metadata piece is added
+     * @param metadataEntry the {@link Entry} to encode
+     */
+    public static void encodeIncrementally(ByteBufAllocator allocator, CompositeByteBuf compositeMetadataBuffer, Entry metadataEntry) {
```

Comment:
We shouldn't need an Entry object to create an entry. This should be done with a method so we don't need to allocate an object.

Reply:
i.e.
encodeIncrementally(ByteBufAllocator allocator, CompositeByteBuf compositeMetadataBuffer, String key, String value)

encodeIncrementally(ByteBufAllocator allocator, CompositeByteBuf compositeMetadataBuffer, WellKnownMimeTypes type, String value)

and so on